### ISSUE TITLE
application or triplestore authentication

### ISSUE THREAD

**Issue by turbomam** [usc-isi-i2/Web-Karma#357]
We anticipate using Karma to convert hospital records into triples, so we need to establish really good security.

We are planning or running the Karma application on a centralized Tomcat or Jetty server, as opposed to requiring all of our collaborators to install it on their own machines.  **Is there any possibility to integrate a login mechanism into Karma itself** (as opposed to application-server security or network security/VPN)?

We also configure Karma to use a GraphDB triplestore as its data/model endpoint.  We would like to password protect our triplestore.  

1. **Can Karma connect to a password-protected RDF4J compliant triple-store for storing models and published triples?**
1. **Can Karma be configured to lock the endpoint text boxes at the bottom of the screen, so a user couldn't redirect the triples to a different server of their choice?**

thanks,
Mark

**Comment by szeke** [comment]
@dkapoor would it be possible for them to stand up an nginx to protect the tomcat?

**Comment by dkapoor** [comment]
Yes, you could set up [nginx](https://www.nginx.com/resources/admin-guide/restricting-access-auth-basic/) to protect tomcat and the Triple Store. They could both me on the same machine and so Karma would be able to write to the Triple store without any additional auth.

Users would need to authenticate themselves when they try to access the Triple Store or Karma.

You can install Karma by source and edit the dialog to disable editing the endpoint textbox. Its a simple HTML/JS change.

**Comment by turbomam** [comment]
thanks, will be discussing with our sysadmins next Friday.

**Comment by dkapoor** [comment]
nginx can be used in front to help with that. Any utility that is not on the same server would then need to enter the credentials. If Karma is on the same machine, it can bypass that, but if you want users to enter the credentials before they publish triples, you will have to add that support in Karma. The RDF publish dialog would also need to accept the credentials that need to be passed to the `PublishRDFCommand` and finally used to make the connection in the `TripleStoreUtil`
